### PR TITLE
Trophy update

### PR DESCRIPTION
**Pull request by Farseer2**
So, my bad on making lots of changes and pushing at once, but still:
- Added a bit of docs to trophies area
- Implemented sceNpTrophyGetGameProgress, sceNpTrophyGetGameIcon and sceNpTrophyGetTrophyIcon
- Moved more memory protection calls to the new api
- Added checksum check of TROPHY.TRP files on version 2+
- Other minor changes

Apparently fixes Splatterhouse, among others

### CODE REVIEW

File: rpcs3/Emu/Cell/PPUModule.cpp [hunk 1]
```diff
@@ -349,7 +350,7 @@ static void ppu_initialize_modules()
 	}
 
 	// Set memory protection to read-only
-	vm::page_protect(ppu_function_manager::addr, ::align(::size32(hle_funcs) * 8, 0x1000), 0, 0, vm::page_writable);
+	utils::memory_protect(vm::base(ppu_function_manager::addr), ::size32(hle_funcs) * 8, utils::protection::ro);
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPUModule.cpp]: It's not a new API for this purpose, it's just different. Memory under the vm management stores its own page information.

**Comment by Zangetsu38** [comment]
Splatterhouse no change
>·W 0:01:10.511425 {PPU[0x1000010] Thread (2023bec0) [0x012999c8]} sceNpTrophy: sceNpTrophyGetTrophyInfo(context=0x1, handle=0x1, trophyId=1, details=*0x0, data=*0xd00cbac0)
·F 0:01:10.511687 {PPU[0x1000010] Thread (2023bec0) [0x012999c8]} MEM: Access violation writing location 0x8

**Comment by Farseer2** [comment]
Will commit the fixes after I test them tonight.
@Zangetsu38 Also added some fixes for sceNpTrophyGetTrophyInfo, so hopefully that would solve that

**Comment by SakataGintokiYT** [comment]
@Farseer2 
Can you check this ?
**[Dragon Ball Z Ultimate Tenkaichi]**
```
U {PPU[0x1000014] Thread (DbPrize thread) [0x00917970]} sceNpTrophy TODO: sceNpTrophyGetRequiredDiskSpace(context=0x1, handle=0x1, reqspace=*0x2cfc4058, options=0x0)
E {PPU[0x1000014] Thread (DbPrize thread) [0x00917958]} sceNpTrophy: sceNpTrophyRegisterContext(context=0x1, handle=0x1, statusCb=*0x770160, arg=*0x2cfc3ca0, options=0x1)
E {PPU[0x1000014] Thread (DbPrize thread) [0x009179b0]} sceNpTrophy: sceNpTrophyGetTrophyUnlockState(context=0x1, handle=0x1, flags=*0x2cfc401c, count=*0xd0177d20)
F {PPU[0x1000000] Thread (main_thread) [0x005fc884]} MEM: Access violation reading location 0x8e02f2d8
```
**[Log:](https://gist.github.com/SakataGintokiYT/ead61e6aa48cb2d5c90604926d571548)**

**Comment by Farseer2** [comment]
@Zangetsu38 Couldn't find it, so I couldn't test the fix, but I fixed the bug that caused it in the new commit (In GetTrophyInfo)
@SakataGintokiYT Will check it out later

**Comment by Zangetsu38** [comment]
yes thx, is fixed now :)

**Comment by JackONeill12** [comment]
Kingdom Hearts 2.5 still has the following error:
E {PPU[0x1000003] Thread (TrophyThread) [0x002269b0]} sceNpTrophy: sceNpTrophyGetTrophyUnlockState(context=0x1, handle=0x1, flags=*0xd002bd24, count=*0xd002bd20)

https://pastebin.com/aND0iPZk

**Comment by Farseer2** [comment]
@JackONeil12 Thats not really an error, just a log

**Comment by JackONeill12** [comment]
Oh ok. My bad.

**Comment by SakataGintokiYT** [comment]
@Farseer2 
[f1483a0f]
**AQUAPAZZA**
This game has a problem with loading trophy data.
```
Fatal error
---------------------------
Unhandled Win32 exception 0xC0000005.
Segfault reading location 0000000000000050 at 00000000002fc854.
Instruction address: 00000000002fc854.
Function address: 00000000002fc7e0 (base+0x2ec7e0).
Module name: 'rpcs3fix.exe'.
Module base: 0000000000010000.
RPCS3 image base: 0000000000010000.
```
`E {PPU[0x1000021] Thread (uTROP:REGIST) [0x003da958]} sceNpTrophy: sceNpTrophyRegisterContext(context=0x1, handle=0x1, statusCb=*0x3edd08, arg=*0x0, options=0x0)`

```
Failed to load trophy data

E {PPU[0x1000000] Thread (main_thread) [0x003d8d30]} cellSysutil: The game cannot continue.
Please end the game.
```
**[Log:](https://gist.github.com/SakataGintokiYT/d6a4c52669dc9cc3af695b50c8379a6e)**

**Comment by Farseer2** [comment]
@SakataGintokiYT Is that crashing since the patch, or has it always crashed?

**Comment by SakataGintokiYT** [comment]
always

File: rpcs3/Loader/TRP.cpp [hunk 2]
```diff
@@ -58,6 +59,40 @@ bool TRPLoader::LoadHeader(bool show)
 		LOG_NOTICE(LOADER, "TRP version: 0x%x", m_header.trp_version);
 	}
 
+	if (m_header.trp_version >= 2)
+	{
+		do
+		{
+			unsigned char hash[20];
```
Comment (paulsapps) [inline rpcs3/Loader/TRP.cpp]: would be much better using std::vector<unsigned char> here

File: rpcs3/Loader/TRP.cpp [hunk 3]
```diff
@@ -58,6 +59,40 @@ bool TRPLoader::LoadHeader(bool show)
...
+			char * file_contents = new char[m_header.trp_file_size];
```
Comment (paulsapps) [inline rpcs3/Loader/TRP.cpp]: would be much better using std::vector here

**Comment by SakataGintokiYT** [comment]
@Farseer2 
Do you have some fix for sceNpTrophyRegisterContext ?

**Comment by Farseer2** [comment]
@SakataGintokiYT Not yet, can't figure out what's wrong with it... But it seems that Thomas was Alone also crashes because of it. Looking into it

**Pull request merged**